feat: introduce XcodeProjMapper to map XcodeProj to XcodeGraph

ajkolean opened the pull request:
**Description:**  
This pull request adds new functionality to convert `.xcworkspace` and `.xcodeproj` files into a high-level graph model (`XcodeGraph`). It introduces:

1. **`XcodeProjToGraph` Module**  
   A domain-focused module responsible for mapping raw Xcode structures (projects, workspaces, schemes, targets, and dependencies) into a cohesive `XcodeGraph` representation. This includes:

   - **Mappers:**  
     - **WorkspaceMapper:** Maps `.xcworkspace` contents into a `Workspace` domain model, identifying referenced `.xcodeproj` files and shared schemes.  
     - **ProjectMapper & PackageMapper:** Extracts projects and associated packages, attributes, and configuration settings, ensuring each project is represented consistently.  
     - **TargetMapper & SchemeMapper:** Converts PBX target definitions and scheme configurations into `Target` and `Scheme` domain entities, capturing essential build, test, and run information.  
     - **DependencyMapper & FileDependencyMapper:** Translates XcodeProj target dependencies, platform conditions, and file references into a structured graph of interdependent components.  
     - **BuildPhaseMapper & BuildRuleMapper:** Interprets build phases and rules, relating source files, resources, and frameworks to their respective domain-level constructs.  
     - **SettingsMapper & ConfigurationMatcher:** Analyzes build settings and configurations, normalizing them into a predictable model that supports further graph analysis or code generation steps.

   - **Extensions & Helpers:**  
     Additional utilities to bridge the gap between raw XcodeProj objects and the domain model. For example, `AbsolutePath+Extensions` helps resolve file references, and `Platform+Extensions` assists in correctly inferring the platform from PBXTarget attributes.

   - **ProcessRunner & LipoTool:**  
     A utility to run external processes (like `lipo`) and integrate their output—such as architecture detection—into the graph model. This allows for richer metadata about binary artifacts and supported architectures.

2. **`TestSupport` Module**  
   This module centralizes mocks, helpers, and utilities that streamline testing. By providing reusable mock objects (e.g., mock project providers, file structures, and build settings), it reduces boilerplate and fosters consistency across all test suites. Key highlights:

   - **Mocks Folder:** A collection of ready-to-use mocks for projects, targets, schemes, environment variables, and more.  
   - **FileCreator & Structure Utilities:** Simplifies creating and handling temporary files or directory structures during tests, ensuring clean and isolated test scenarios.

3. **`XcodeProjToGraphTests`**  
   A comprehensive test suite validating each piece of the mapping logic. The tests are thoughtfully organized by domain:

   - **MapperTests:**  
     - **Build, Dependency, Settings, Target, Workspace:** Ensure that each mapper converts raw Xcode structures (phases, dependencies, settings, etc.) into domain-level models accurately.
     - **Project & Package:** Validate correct interpretation of project attributes, packages, and configuration lists.
     - **Scheme & Graph:** Confirm that schemes are parsed and integrated into the workspace and project graphs, and that the overall graph mapper works as intended.

   - **ParserTests (ProjectParserTests):**  
     Verifies that the parsing logic for `.xcodeproj` and `.xcworkspace` directories is robust and reliable, providing a consistent foundation for the mappers.

   - **ProcessTests (LipoToolTests, ProcessRunnerTests):**  
     Checks the integration with external tools (`lipo`) and process running capabilities, ensuring errors are handled gracefully and output is captured correctly.

**Next Steps - Integration Tests:**  
While this PR focuses on unit-level validation, an upcoming PR will introduce integration tests that run end-to-end scenarios using real `.xcworkspace` and `.xcodeproj` fixtures. This will confirm that the entire pipeline—from reading disk-based Xcode structures to producing a final `XcodeGraph`—functions correctly in practice.

While this PR focuses on core functionality, the integration tests validating end-to-end scenarios (e.g., parsing `.xcworkspace` and `.xcodeproj` into `XcodeGraph`) are complete. However, due to the size of the test fixtures (15,000 lines), they were removed to reduce noise.

To review and run these tests, check out this commit (before they were removed) : 75bf656adcec3b5fe7f4c034f797961361f9f584. This includes real-world project fixtures and ensures the entire pipeline functions as expected in practical scenarios.

The results for these are browsable here: https://github.com/tuist/XcodeGraph/blob/75bf656adcec3b5fe7f4c034f797961361f9f584/Tests/XcodeProjToGraphTests/IntegrationTests/Projects/IntegrationTests%2BcommandLineToolWithDynamicFramework.swift

**Testing & Verification:**  
- **Locally:**  
  - Run `swift build` to compile all modules.
  - Run `swift test` to execute the full test suite. All tests should pass, confirming correct functionality.
- **CI Integration:**  
  Once integrated with CI (e.g., GitHub Actions), these tests will run automatically on each PR or commit, ensuring ongoing stability and reliability.

**No Breaking Changes:**  
This PR adds new modules and tests without altering existing functionality. It forms the foundation for further enhancements and refinements.

**Documentation & Maintenance:**  
Code comments and some documentation are included to guide future developers. With a clear architecture, it will be easier to extend capabilities, add integration tests, and enhance performance or features over time.

---

**Summary:**  
This PR significantly expands the repository’s capabilities by introducing `XcodeProjToGraph` and `TestSupport` modules, along with a robust test suite in `XcodeProjToGraphTests`. Together, they provide a structured path from raw Xcode project/workspace data to a domain-level `XcodeGraph` model. The next step will be to add integration tests to validate complex real-world scenarios, ensuring the tool’s practicality and reliability for future use cases.

fortmarek commented on Sources/XcodeProjToGraph/Mappers/TargetAndSchemes/TargetMapper.swift:
```diff
@@ -0,0 +1,285 @@
+
+    // MARK: - Helper Methods
+
+    private func extractBundleIdentifier(from target: PBXTarget) throws -> String {
+        if let bundleId = target.debugBuildSettings.string(for: .productBundleIdentifier) {
+            return bundleId
+        }
+        throw MappingError.missingBundleIdentifier(targetName: target.name)
+    }
```
What do you think about doing this instead:
```swift
extension PBXTarget {
  fileprivate func bundleIdentifier() throws -> String {
     ...
  }
}
```

fortmarek commented on Sources/XcodeProjToGraph/ProcessRunner/Executable.swift:
Let's use a utility that we built for running shell commands instead: https://github.com/tuist/command

fortmarek commented on Sources/XcodeProjToGraph/ProcessRunner/Lipo/LipoArchsResult.swift:
```diff
@@ -0,0 +1,38 @@
+/// - Throws:
+///   - `ProcessRunnerError.failedToRunProcess` if `lipo` exits with a non-zero code.
+///   - Any errors related to decoding or interpreting the output as architectures.
+public func parseLipoArchsResult(_ result: ProcessResult) throws -> LipoArchsResult {
```
Have you considered using the implementation that we have in tuist: https://github.com/tuist/tuist/blob/main/Sources/TuistCore/MetadataProviders/FrameworkMetadataProvider.swift ?

I believe we can use it instead. If so, then I'd consider creating a new module called something like `ProductMetadataProvider` or `BinaryMetadataProvider` that we can then share back with `tuist` instead of having the implementation duplicated. If we end up doing this, I'd consider moving the whole `MetadataProviders` directory into `XcodeGraph`: https://github.com/tuist/tuist/tree/main/Sources/TuistCore/MetadataProviders

ajkolean replied:
That’s a great callout. Thanks for pointing this out! I’ve added TODO comments to explore how best to align with this or where it should live. It might not be as straightforward as dropping it in, since the `FileHandler` here already uses `TuistSupport` (though only for the Info.plist). That said, I think this could be a great opportunity to modularize FileHandling or maybe more as a whole, especially if we can align with the broader patterns in Tuist.

fortmarek replied:
The `FileHandler` should be removed in favor of `FileSystem`. It should be already possible. If that's the only blocking thing to move it here, then we should be good.

fortmarek commented on Tests/XcodeProjToGraphTests/MapperTests/Xcode/Build/BuildRuleMapperTests.swift:
```diff
@@ -0,0 +1,134 @@
+@testable import TestSupport
+@testable import XcodeProjToGraph
+
+@Suite
+struct BuildRuleMapperTests {
```
Yes to Swift testing 👏

fortmarek reviewed (commented):
First of all, thanks a lot @ajkolean 👏 

This work is huge and will enable powerful workflows that the Swift community has not seen, yet.

I did a first pass, trying to mostly focus on the higher level structure of the codebase.

Some of the comments are applicable across the PR, so it'd be amazing if you could have that in mind as I didn't want to repeat myself too much. I tried to point out when I saw some pattern we should imho redo in other places as well, but I might not have done this perfectly.

Overall, I am very onboard with the structure of breaking down the mapping of `XcodeProj` into multiple layers of individual mappers that have a simple responsibility.

Some of the patterns I'd consider changing (and these are also captured in the individual review comments):
- Prefer structs over classes
- Minimize the public interface of the `XcodeProjToGraph` module to the bare minimum. This will help us keep this repository stable.
- Don't overdo it with running things in parallel. Benchmark first and make asynchronous only parts of the code that are actually resource intensive.
- Use our utility for running shell commands instead of having a new implementation: https://github.com/tuist/command
- Utilities should usually be configured directly via methods instead of passing the configuration at the `init` time. Unless there's a clear win by configuration utilities through `init`, let's keep `init` reserved for dependency injection.
- Individual mappers should not be accessing layers above. For example, a `PBXBuildPhase` mapper should not have a notion of `PBXTarget`.

Let me know if we can help you in any way to proceed. This is top on my priority list, so I'll try to provide feedback swiftly here as you go through the review.

Once higher-level concerns are resolved, I'll do another pass going deeper into the individual implementations as I skimmed through some of them during the initial pass.

This must have been a ton of work and we really appreciate it – this is open source at its best 💜

ajkolean commented on Package.swift:
```diff
@@ -13,6 +13,37 @@ var targets: [Target] = [
         ]
     ),
+    .target(
+        name: "XcodeProjToGraph",
```
Renamed to `XcodeProjMapper` to better reflect its purpose

ajkolean commented on Package.swift:
```diff
@@ -13,6 +13,37 @@ var targets: [Target] = [
+            "XcodeProjToGraph",
+        ],
+        path: "Sources/TestSupport"
+    ),
```
Integrated the mocks directly into XcodeProjMapperTests, removing the separate TestSupport target

ajkolean commented on Sources/TestSupport/Mocks/MockBuildFilesAndPhases.swift:
```diff
@@ -0,0 +1,139 @@
+import Foundation
+import Path
+import XcodeGraph
+@testable @preconcurrency import XcodeProj
```
After reverting to synchronous code and removing concurrency bridging, it’s no longer necessary and removed.

ajkolean commented on Sources/TestSupport/Mocks/MockBuildFilesAndPhases.swift:
Reorganized test data into TestData files per model

ajkolean commented on Sources/TestSupport/Mocks/MockBuildFilesAndPhases.swift:
```diff
@@ -0,0 +1,139 @@
+import XcodeGraph
+@testable @preconcurrency import XcodeProj
+
+extension PBXBuildFile {
```
Split the test data into multiple files

ajkolean commented on Sources/TestSupport/Mocks/MockBuildFilesAndPhases.swift:
```diff
@@ -0,0 +1,139 @@
+        let buildFile = PBXBuildFile(file: file, settings: settings)
+        pbxProj.add(object: buildFile)
+        return buildFile
+    }
```
I simplified mocks to return just the PBXBuildFile and not introduce hidden side effects

ajkolean commented on Sources/TestSupport/Mocks/MockProjectProvider.swift:
```diff
@@ -0,0 +1,97 @@
+
+@testable import XcodeProjToGraph
+
+public struct MockWorkspaceProvider: WorkspaceProviding {
```
That’s really cool and definitely a time saver! I wonder if it’s the right fit here, though. Since we only have a few simple mocks, adding a dependency with swift-syntax introduces more complexity and significant build-time overhead than needed. That said, I’m not against it. Just want to make sure we’re fully aware of the trade-offs.

ajkolean commented on Sources/XcodeProjToGraph/Extensions/Package+Extensions.swift:
```diff
@@ -0,0 +1,13 @@
+        case let .local(path):
+            return path.pathString
+        }
+    }
```
`ProjectParser` is the only public one now

ajkolean commented on Sources/XcodeProjToGraph/Mappers/Build/BuildPhaseMapper.swift:
```diff
@@ -0,0 +1,428 @@
+
+    public func mapSources(target: PBXTarget) async throws -> [SourceFile] {
+        guard let sourcesPhase = try target.sourcesBuildPhase() else { return [] }
+        return try await sourcesPhase.files?.asyncCompactMap { try await self.mapSourceFile($0) }
```
💯 I ran some comparisons and found that running the Tuist repo's Xcode workspace with the old async implementation took about 2 seconds, while the sync version took around 5 seconds. However, for smaller projects, the overhead of async/await made the sync version noticeably faster. Given this, I’ve removed the async code for now, as it does feel like it might have been a case of premature optimization

fortmarek replied:
Let's revisit this later on – those numbers are not nothing, so I think we should consider some parallelization. But let's first get a stable synchronous version going and then we can look into parallelizing the parts where it would make a big difference.

ajkolean commented on Sources/XcodeProjToGraph/Mappers/Graph/GraphMapper.swift:
```diff
@@ -0,0 +1,169 @@
+            }
+
+        // Map each project in the workspace
+        let projectResults = try await workspace.projects.lazy.asyncCompactMap { path in
```
Removed I did out of habit

ajkolean commented:
@fortmarek 
Thanks for the detailed feedback! I’m guessing waking up to what probably looked like a 7,000-line horror show wasn’t exactly fun. 🙈 I really appreciate you diving into it so thoroughly and quickly, especially with everything else on your plate, it means a lot.

I’ve gone through and addressed almost everything: cleaned up the mappers, streamlined the test data, and tightened up public access. If there’s anything I missed or further tweaks you’d suggest, I’m all ears.

Thanks again for taking the time to go through this so thoughtfully. Your sharp eye and collaborative approach made this a great process. Looking forward to any more thoughts! 🙏

fortmarek commented on Sources/XcodeProjMapper/Mappers/Targets/PBXTarget+PlatformInference.swift:
```diff
@@ -0,0 +1,85 @@
+            result.formUnion(Platform.visionOS.destinations)
+        }
+
+        guard result.isEmpty else { return result }
+
+        // Fallback if no platform detected.
+        let product = productType?.mapProductType() ?? .app
```
Should we consider throwing instead of falling back to `.app`?

fortmarek commented on Sources/XcodeProjMapper/Mappers/Targets/PBXTargetDependencyMapper.swift:
```diff
@@ -0,0 +1,129 @@
+            }
+        }
+
+        return nil
```
Should we throw instead?

fortmarek commented on Sources/XcodeProjMapper/Mappers/Targets/PBXTargetDependencyMapper.swift:
```diff
@@ -0,0 +1,129 @@
+            // Direct reference to another target in the same project.
+            return .target(name: remoteInfo, status: .required, condition: condition)
+        case let .fileReference(fileReference):
+            guard let projectRelativePath = fileReference.path else { return nil }
```
I'd consider throwing here.

fortmarek commented on Sources/XcodeProjMapper/Mappers/Targets/PBXTargetDependencyMapper.swift:
```diff
@@ -0,0 +1,129 @@
+            // Reference to a target in another project.
+            return .project(target: remoteInfo, path: absPath, status: .required, condition: condition)
+        case .unknownObject:
+            return nil
```
I'd also throw here.

fortmarek commented on Sources/XcodeProjMapper/Mappers/Targets/PBXTargetMapper.swift:
```diff
@@ -0,0 +1,312 @@
+        )
+    }
+
+    // MARK: - Helpers
```
Let's mark the helpers as `private`

fortmarek reviewed (commented):
Some additional feedback. Mostly nits. Some bigger pieces of feedback from this round I'd like to call out:
- When we run into unknown objects and scenarios, I'd lean into throwing instead of swallowing these objects as `nil`. That will make it easier for us to spot our blinds spots. Otherwise, the conversion might be lossier than one would expect and then it's gonna be hard to figure out why.
- I'd consider removing `ProjectParser` and making `GraphMapper` as the public interface. To do that, we'll need to make a PR in `XcodeProj` to accept an `AbsolutePath` in its `init`
- Let's also update `XcodeProj` to hold its own `path` – I'd like us to move away from the `ProjectProvider` and `WorkspaceProvider` helpers and use `XcodeProj` directly.

fortmarek commented on Tests/XcodeProjMapperTests/IntegrationTests/IntegrationTests.swift:
```diff
@@ -0,0 +1,173 @@
+        record: .failed
+    )
+)
+struct IntegrationTests {
+    /// Asserts that the given graph parsed from a workspace fixture matches the provided inline snapshots.
+    ///
+    /// - Parameters:
+    ///   - fixture: A closure that returns a `WorkspaceFixture` used to generate the graph.
```
Why are we asserting with snapshots? I'd instead compare the parsed `XcodeGraph` with the expected `XcodeGraph`:
```swift
#expect(gotXcodeGraph, expectedXcodeGraph)
```

What's the reason we went with asserting on snapshot text which is more unstable than comparing the `XcodeGraph` directly?

fortmarek commented on Sources/XcodeProjMapper/Mappers/Workspace/XCWorkspaceMapper.swift:
```diff
@@ -0,0 +1,125 @@
+        from srcPath: AbsolutePath,
+        workspaceProvider: WorkspaceProviding
+    ) throws -> [Scheme] {
+        var schemes = [Scheme]()
+        let sharedDataPath = Path(srcPath.pathString) + "xcshareddata/xcschemes"
+
+        if sharedDataPath.exists {
+            let schemePaths = try sharedDataPath.children().filter { $0.extension == "xcscheme" }
+
+            // Construct graphType for schemes
+            let graphType = GraphType.workspace(workspaceProvider)
+            let schemeMapper = XCSchemeMapper()
+            for schemePath in schemePaths {
+                let xcscheme = try XCScheme(path: schemePath)
+                let scheme = try schemeMapper.map(xcscheme, shared: true, graphType: graphType)
+                schemes.append(scheme)
+            }
+        }
+
+        return schemes
```
nit: I generally prefer not using `var`s if not necessary as it makes it easier to spot where the final values are coming from:
```suggestion
        let sharedDataPath = Path(srcPath.pathString) + "xcshareddata/xcschemes"

        if sharedDataPath.exists {
            let schemePaths = try sharedDataPath.children().filter { $0.extension == "xcscheme" }

            try schemePaths.map { schemePath in
                try schemeMapper.map(
                  try XCScheme(path: schemePath), 
                  shared: true, 
                  graphType: .workspace(workspaceProvider)
                )
            }
        } else {
          return []
        }
```

Note I removed `schemeMapper` initialization as that belongs into the `init` of `XCWorkspaceMapper` as it should be dependency injected

ajkolean commented:
Apologies for the delay. I was away due to the holidays. I will respond and update from the feedback within the next few days. I haven't forgotten about it.

fortmarek commented:
> Apologies for the delay. I was away due to the holidays. I will respond and update from the feedback within the next few days. I haven't forgotten about it.

No worries, I figured 🙂 Hope you were able to recharge over the holidays! Looking forward to updates on this, we are still very much excited about this work landing.

fortmarek commented on Sources/XcodeProjMapper/Mappers/Graph/XcodeGraphMapper.swift:
```diff
@@ -0,0 +1,291 @@
+            projects: projectPaths
+        )
+
+        // Step 3: For each project path, load the .xcodeproj and map to a `Project`
+        let projectResults = try workspace.projects.map { path -> (AbsolutePath, Project) in
+            // Instead of "ProjectProvider", just load XcodeProj directly
+            let xcodeProj = try XcodeProj(pathString: path.pathString)
+            let projectMapper = PBXProjectMapper()
+            let project = try projectMapper.map(xcodeProj: xcodeProj)
+            return (path, project)
+        }
```
What do you think about turning some of these more complex steps into helper methods where the name of the method would clearly indicate what it does?

fortmarek commented on Sources/XcodeProjMapper/Mappers/Packages/XCPackageMapper.swift:
```diff
@@ -0,0 +1,83 @@
+    ///
+    /// - Parameter package: The remote package reference containing the version requirement.
+    /// - Returns: A `Package.Requirement` reflecting the specified versioning scheme.
+    func mapRequirement(package: XCRemoteSwiftPackageReference) -> Requirement {
```
Ping on this

fortmarek commented on Sources/XcodeProjMapper/Mappers/Phases/PBXScriptsBuildPhaseMapper.swift:
```diff
@@ -0,0 +1,99 @@
+    func map(
+        _ scriptPhases: [PBXShellScriptBuildPhase],
+        buildPhases: [PBXBuildPhase],
+        xcodeProj _: XcodeProj
```
doesn't seem to be used.

fortmarek commented on Sources/XcodeProjMapper/Mappers/Phases/PBXSourcesBuildPhaseMapper.swift:
```diff
@@ -0,0 +1,52 @@
+
+struct PBXSourcesBuildPhaseMapper: PBXSourcesBuildPhaseMapping {
+    func map(_ sourcesBuildPhase: PBXSourcesBuildPhase, projectProvider: ProjectProviding) throws -> [SourceFile] {
+        guard let files = sourcesBuildPhase.files, !files.isEmpty else {
+            return []
+        }
```
Ping on this.

fortmarek commented on Sources/XcodeProjMapper/Mappers/Settings/XCConfigurationList+Helpers.swift:
```diff
@@ -0,0 +1,36 @@
+    ///
+    /// - Parameter key: The `BuildSettingKey` to look up.
+    /// - Returns: The value as a `String` if found, otherwise `nil`.
+    func stringSetting(for key: BuildSettingKey) -> String? {
+        for config in buildConfigurations {
+            if let value = config.buildSettings.string(for: key) {
+                return value
+            }
+        }
+        return nil
+    }
```
Ping on this.

fortmarek commented on Sources/XcodeProjMapper/Mappers/Targets/PBXTarget+BuildSettings.swift:
```diff
@@ -0,0 +1,80 @@
+    /// Retrieves the path to the Info.plist file from the target's build settings.
+    ///
+    /// - Returns: The `INFOPLIST_FILE` value if present, otherwise `nil`.
+    func infoPlistPath() throws -> String? {
+        buildConfigurationList?.stringSetting(for: .infoPlistFile)
+    }
+
+    /// Retrieves the path to the entitlements file from the target's build settings.
+    ///
+    /// - Returns: The `CODE_SIGN_ENTITLEMENTS` value if present, otherwise `nil`.
+    func entitlementsPath() throws -> String? {
+        buildConfigurationList?.stringSetting(for: .codeSignEntitlements)
+    }
+
+    /// Retrieves deployment target versions for various platforms supported by this target.
+    ///
+    /// Checks build configurations for:
+    /// - `IPHONEOS_DEPLOYMENT_TARGET`
+    /// - `MACOSX_DEPLOYMENT_TARGET`
+    /// - `WATCHOS_DEPLOYMENT_TARGET`
+    /// - `TVOS_DEPLOYMENT_TARGET`
+    /// - `VISIONOS_DEPLOYMENT_TARGET`
+    ///
+    /// - Returns: A `DeploymentTargets` instance containing any discovered versions.
+    func deploymentTargets() throws -> DeploymentTargets {
```
Ping on this

fortmarek commented on Sources/XcodeProjMapper/Utilities/DeveloperDirectoryProvider.swift:
```diff
@@ -0,0 +1,26 @@
+    func developerDirectory() async throws -> AbsolutePath
+}
+
+public struct DeveloperDirectoryProvider: DeveloperDirectoryProviding {
```
I wouldn't make this public as it's not expected that `XcodeProjMapper` would suddenly be a utility module for services like this one.

fortmarek commented on Tests/MetadataProvidersTests/FrameworkMetadataProviderTests.swift:
Is there a reason why the tests here are commented out?

fortmarek commented on Tests/XcodeProjMapperTests/Mocks/MockProjectProvider.swift:
This should be no longer needed?

fortmarek commented on Tests/XcodeProjMapperTests/Mocks/PBXProjectMapper+Mock.swift:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    /// Creates a mapped graph from multiple project providers, useful for testing multi-project scenarios.
+//    func createMappedGraph(
+//        graphType: XcodeMapperGraphType,
+//        projectProviders: [AbsolutePath: MockProjectProvider]
+//    ) throws -> XcodeGraph.Graph {
+//        let mapper = XcodeGraphMapper() { path in
+//            guard let provider = projectProviders[path] else {
+//                Issue.record("Unexpected project path requested: \(path)")
+//                throw XcodeProjMapper.XcodeProjError.noProjectsFound
+//            }
+//            return provider
+//        }
+//
+//        return try mapper.map()
+//    }
```
Can we remove this?

fortmarek commented on Tests/XcodeProjMapperTests/Mocks/WorkspaceFixture.swift:
```diff
@@ -0,0 +1,68 @@
+
+    // NOTE: - This is temporary to reduce PR noise
+    /// Unzips and returns path to the Fixtures directory, if not already done.
+    static func getFixturesPath() throws -> String {
```
Ping on this.

fortmarek commented on Tests/XcodeProjMapperTests/TestData/.swift:
This file can be removed?

fortmarek commented on Tests/XcodeProjMapperTests/TestData/XcodeProj+TestData.swift:
```diff
@@ -0,0 +1,25 @@
+        ),
+        mainGroup: PBXGroup,
+        targets: [PBXTarget] = [PBXNativeTarget.test()],
+        schemes _: [XCScheme] = []
```
Ping on this.

fortmarek reviewed (commented):
Thanks for this next iteration! I have another round of comments, but I feel we're very close to merging this 🙂

ajkolean commented:
> Thanks for this next iteration! I have another round of comments, but I feel we're very close to merging this 🙂

Thanks for another round! 😅 I didn't realize I pushed these commits yet. Of course, any feedback is always appreciated, but I was planning to address or comment on anything unresolved before asking for another look. I’ll tidy things up and give you a shout when it’s fully ready! 🙏 I think most of the moving pieces are in place now.

ajkolean commented:
Hey @fortmarek! Thanks for all the thorough feedback. I’ve just pushed the latest changes that (hopefully) address the remaining points. Apologies if I missed anything; please let me know if you spot any lingering issues or have more suggestions. 😬

fortmarek commented on Sources/XcodeProjMapper/Extensions/XCWorkspace+Extensions.swift:
```diff
@@ -0,0 +1,27 @@
+
+extension XCWorkspace {
+    /// A computed property that either returns the workspace’s `path`
+    /// or throws a `MissingWorkspacePathError` if it’s `nil`.
```
```suggestion
```

We crash if that happens, so we can remove that part of the comment.

fortmarek commented on Sources/XcodeProjMapper/Extensions/XCWorkspace+Extensions.swift:
```diff
@@ -0,0 +1,27 @@
+
+extension XcodeProj {
+    /// A computed property that either returns the project’s `path`
+    /// or throws a `MissingProjectPathError` if it’s `nil`.
```
```suggestion
```

fortmarek commented on Sources/XcodeProjMapper/Mappers/Graph/XcodeGraphMapper.swift:
```diff
@@ -0,0 +1,282 @@
+
+    // MARK: - Public API
+
+    public func map(at path: AbsolutePath) async throws -> Graph {
```
I'd put this function also in the `XcodeGraphMapping` protocol.

ajkolean replied:
It is in the protocol. I updated the internal parameter name of the protocol to match. Must of missed it when updating and compiler doesn't care if internal ones match

fortmarek reviewed (approved):
This is huge @ajkolean. I have just a couple of last nits after which we can finally merge this 🚀 I am _very_ excited about it 🙂

ajkolean commented on Sources/XcodeGraph/Models/Plist.swift:
```diff
@@ -92,26 +92,31 @@ extension Dictionary where Value == Plist.Value {
 
 // MARK: - InfoPlist
 
+public struct infoPlist: Equatable, Codable, Sendable {
```
Whoops, this was an oversight, I was trying to decide between wrapper struct or include configs as associated values. I removed this, good catch!

ajkolean commented:
A few notes/reminders on somethings:

- I’ve removed integration and performance tests for now. These can easily be added in a follow-up PR. To keep this PR focused and manageable, I opted not to introduce the fixtures at this stage.
- This PR extracted some Metadata Providers from the main repo. We’ll need to ensure the main repo is updated accordingly to reflect this change.

fortmarek commented:
> I’ve removed integration and performance tests for now. These can easily be added in a follow-up PR. To keep this PR focused and manageable, I opted not to introduce the fixtures at this stage.

I'm aware of that, that's fine. Performance testing can be done a bit more ad-hoc instead of having these automated. We have a [fixture generator](https://github.com/tuist/tuist/blob/main/.mise/tasks/fixturegenerator/run) in the `tuist/tuist` repository to generate arbitrarily large Xcode projects.

> This PR extracted some Metadata Providers from the main repo. We’ll need to ensure the main repo is updated accordingly to reflect this change.

Yes! Would you be up to take that work? If not, I could try to carve out some time for it as well.

fortmarek commented:
@ajkolean would you have some time to look into the CI failures? Those look legit.

ajkolean commented:
>Yes! Would you be up to take that work? If not, I could try to carve out some time for it as well.

Sure I can update after this in.

> @ajkolean would you have some time to look into the CI failures? Those look legit.

Updated. I had to bump the macOS version and swift version due to some of the dependencies:

`Compiling for macOS 12.0, but module 'FileSystem' has a minimum deployment target of macOS 13.0`
`Dependencies could not be resolved because 'command' >= 0.2.1 contains incompatible tools version (5.10.0)`

fortmarek commented:
Great work @ajkolean 👏 Let's iterate on the blogpost to share the news that this is now available with the community 💯
